### PR TITLE
[Compute] Update API version to 2019-07-01

### PR DESCRIPTION
**Pull request by qwordy**
---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [ ] The PR has modified HISTORY.rst describing any customer-facing, functional changes. Note that this does not include changes only to help content. (see [Modifying change log](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules#modify-change-log)).

- [ ] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

### CODE REVIEW

**Comment by qwordy** [comment]
So many failed test cases due to the new API version.
test_vm_commands.VMShowListSizesListIPAddressesScenarioTest.test_vm_show_list_sizes_list_ip_addresses
test_vm_commands.VMImageWithPlanTest.test_vm_create_with_market_place_image
test_vm_commands.VMAvailSetLiveScenarioTest.test_vm_availset_convert
test_vm_commands.VMDiskAttachDetachTest.test_vm_ultra_ssd_storage_sku
test_vm_commands.VMDiskAttachDetachTest.test_vmss_ultra_ssd_storage_sku
test_vm_commands.MSIScenarioTest.test_vm_explicit_msi
test_vm_commands.MSIScenarioTest.test_vmss_msi
test_vm_commands.VMZoneScenarioTest.test_vmss_create_single_zone
test_vm_commands.MSIScenarioTest.test_vm_msi
test_vm_commands.VMZoneScenarioTest.test_vmss_create_zonal_with_fd
test_vm_commands.MSIScenarioTest.test_vmss_explicit_msi
test_vm_commands.VMZoneScenarioTest.test_vm_create_zones
test_vm_commands.VMDiskEncryptionTest.test_vm_disk_encryption_e2e
test_vm_commands.VMCreateWithExistingNic.test_vm_create_existing_nic
test_vm_commands.VMZoneScenarioTest.test_vmss_create_x_zones
test_vm_commands.VMGalleryImage.test_gallery_e2e

**Comment by qwordy** [comment]
I will comment these test cases and fix them later!
test_vm_commands.VMImageWithPlanTest.test_vm_create_with_market_place_image
test_vm_commands.MSIScenarioTest.test_vm_explicit_msi
test_vm_commands.MSIScenarioTest.test_vm_msi
test_vm_commands.MSIScenarioTest.test_vmss_msi
test_vm_commands.MSIScenarioTest.test_vmss_explicit_msi
test_vm_commands.VMCreateWithExistingNic.test_vm_create_existing_nic

File: src/azure-cli/azure/cli/command_modules/vm/custom.py [hunk 1]
```diff
@@ -2749,7 +2758,17 @@ def create_image_version(cmd, resource_group_name, gallery_name, gallery_image_n
                                             target_regions=target_regions or [TargetRegion(name=location)],
                                             source=source, replica_count=replica_count,
                                             storage_account_type=storage_account_type)
-    image_version = ImageVersion(publishing_profile=profile, location=location, tags=(tags or {}))
+    if cmd.supported_api_version(min_api='2019-07-01', operation_group='gallery_image_versions'):
+        GalleryImageVersionStorageProfile = cmd.get_models('GalleryImageVersionStorageProfile')
+        GalleryArtifactVersionSource = cmd.get_models('GalleryArtifactVersionSource')
+        # GalleryOSDiskImage = cmd.get_models('GalleryOSDiskImage')
```
Comment (MyronFanQiu) [inline src/azure-cli/azure/cli/command_modules/vm/custom.py]: why do we need these two lines?

File: src/azure-cli/azure/cli/command_modules/vm/tests/latest/test_vm_commands.py [hunk 2]
```diff
@@ -103,7 +103,7 @@ def test_vm_show_list_sizes_list_ip_addresses(self, resource_group):
         self.kwargs.update({
             'loc': 'centralus',
             'vm': 'vm-with-public-ip',
-            'allocation': 'dynamic',
+            'allocation': 'static',
```
Comment (MyronFanQiu) [inline src/azure-cli/azure/cli/command_modules/vm/tests/latest/test_vm_commands.py]: intentional? any reason?

File: src/azure-cli/azure/cli/command_modules/vm/tests/latest/test_vm_commands.py [hunk 3]
```diff
@@ -3169,7 +3171,7 @@ def test_vm_create_existing_nic(self, resource_group):
         self.cmd('network vnet create -g {rg} -n my-vnet --subnet-name my-subnet1')
         self.cmd('network nic create -g {rg} -n my-nic --subnet my-subnet1 --vnet-name my-vnet --public-ip-address my-pip')
         self.cmd('network nic ip-config create -n my-ipconfig2 -g {rg} --nic-name my-nic --private-ip-address-version IPv6')
-        self.cmd('vm create -g {rg} -n vm1 --image ubuntults --nics my-nic --generate-ssh-keys --admin-username ubuntuadmin')
+        self.cmd('vm create -g {rg} -n vm1 --image centos --nics my-nic --generate-ssh-keys --admin-username ubuntuadmin')
```
Comment (MyronFanQiu) [inline src/azure-cli/azure/cli/command_modules/vm/tests/latest/test_vm_commands.py]: so, can we create ubuntults right now?

File: src/azure-cli/azure/cli/command_modules/vm/custom.py (hunk 1 above)
Comment (qwordy) [reply src/azure-cli/azure/cli/command_modules/vm/custom.py]: I will add new type of source.

File: src/azure-cli/azure/cli/command_modules/vm/tests/latest/test_vm_commands.py (hunk 2 above)
Comment (qwordy) [reply src/azure-cli/azure/cli/command_modules/vm/tests/latest/test_vm_commands.py]: follow error message's guide

File: src/azure-cli/azure/cli/command_modules/vm/tests/latest/test_vm_commands.py (hunk 3 above)
Comment (qwordy) [reply src/azure-cli/azure/cli/command_modules/vm/tests/latest/test_vm_commands.py]: no

**Comment by qwordy** [comment]
test_sqlvm_commands.SqlVmScenarioTest.test_sqlvm_create_and_delete
test_sqlvm_commands.SqlVmAndGroupScenarioTest.test_sqlvm_add_and_remove

File: src/azure-cli/azure/cli/command_modules/vm/custom.py [hunk 4]
```diff
@@ -262,6 +262,15 @@ def create_managed_disk(cmd, resource_group_name, disk_name, location=None,  # p
     else:
         option = DiskCreateOption.empty
 
+    if source_storage_account_id is None and source_blob_uri is not None:
+        from azure.cli.core.commands.client_factory import get_subscription_id
+        from msrestazure.tools import resource_id
+        subscription_id = get_subscription_id(cmd.cli_ctx)
+        storage_account_name = source_blob_uri.split('.')[0].split('/')[-1]
+        source_storage_account_id = resource_id(
+            subscription=subscription_id, resource_group=resource_group_name,
```
Comment (jiasli) [inline src/azure-cli/azure/cli/command_modules/vm/custom.py]: The subscription ID of the Storage Account may not be the current one. Per [doc](https://docs.microsoft.com/en-us/cli/azure/disk?view=azure-cli-latest#az-disk-create)
```
--source-storage-account-id
Used when source blob is in a different subscription.
```

File: src/azure-cli/azure/cli/command_modules/vm/custom.py [hunk 5]
```diff
@@ -2749,7 +2758,17 @@ def create_image_version(cmd, resource_group_name, gallery_name, gallery_image_n
...
-    image_version = ImageVersion(publishing_profile=profile, location=location, tags=(tags or {}))
+    if cmd.supported_api_version(min_api='2019-07-01', operation_group='gallery_image_versions'):
+        GalleryImageVersionStorageProfile = cmd.get_models('GalleryImageVersionStorageProfile')
+        GalleryArtifactVersionSource = cmd.get_models('GalleryArtifactVersionSource')
+        # GalleryOSDiskImage = cmd.get_models('GalleryOSDiskImage')
+        # GalleryDataDiskImage = cmd.get_models('GalleryDataDiskImage')
+        source = GalleryArtifactVersionSource(id=managed_image)
+        storage_profile = GalleryImageVersionStorageProfile(source=source)
+        image_version = ImageVersion(publishing_profile=profile, location=location, tags=(tags or {}),
+                                     storage_profile=storage_profile)
```
Comment (jiasli) [inline src/azure-cli/azure/cli/command_modules/vm/custom.py]: If `storage_profile` and `profile` both all derived from `managed_image`, one of them seems to be duplicated. I doubt the usage is correct.

File: src/azure-cli/azure/cli/command_modules/vm/custom.py (hunk 4 above)
Comment (qwordy) [reply src/azure-cli/azure/cli/command_modules/vm/custom.py]: User needs to specify it if it's in a different subscription.

File: src/azure-cli/azure/cli/command_modules/vm/custom.py (hunk 5 above)
Comment (qwordy) [reply src/azure-cli/azure/cli/command_modules/vm/custom.py]: They are mandatory

File: src/azure-cli/azure/cli/command_modules/vm/custom.py [hunk 6]
```diff
@@ -262,6 +262,15 @@ def create_managed_disk(cmd, resource_group_name, disk_name, location=None,  # p
     else:
         option = DiskCreateOption.empty
 
+    if source_storage_account_id is None and source_blob_uri is not None:
+        from azure.cli.core.commands.client_factory import get_subscription_id
+        from msrestazure.tools import resource_id
+        subscription_id = get_subscription_id(cmd.cli_ctx)
+        storage_account_name = source_blob_uri.split('.')[0].split('/')[-1]
```
Comment (MyronFanQiu) [inline src/azure-cli/azure/cli/command_modules/vm/custom.py]: Is it able to use regex?

File: src/azure-cli/azure/cli/command_modules/vm/custom.py (hunk 1 above)
Comment (MyronFanQiu) [reply src/azure-cli/azure/cli/command_modules/vm/custom.py]: do you mean in the future?

File: src/azure-cli/azure/cli/command_modules/vm/tests/latest/test_vm_commands.py (hunk 2 above)
Comment (MyronFanQiu) [reply src/azure-cli/azure/cli/command_modules/vm/tests/latest/test_vm_commands.py]: so, user cannot use dynamic allocation?

File: src/azure-cli/azure/cli/command_modules/vm/tests/latest/test_vm_commands.py [hunk 7]
```diff
@@ -1690,21 +1692,21 @@ def test_vm_disk_create_with_standard_zrs_sku(self, resource_group):
...
+    @ResourceGroupPreparer(name_prefix='cli-test-ultrassd', location='eastus')
     def test_vmss_ultra_ssd_storage_sku(self, resource_group):
 
         self.kwargs.update({
             'vmss': 'vm-ultrassd',
             'vmss2': 'vm-ultrassd2'
         })
         self.cmd('vmss create -g {rg} -n {vmss} --admin-username admin123 --admin-password testPassword0 --image debian --storage-sku UltraSSD_LRS '
-                 ' --data-disk-sizes-gb 4 --zone 2 --location eastus2 --vm-sku Standard_D2s_v3')
+                 ' --data-disk-sizes-gb 4 --zone 2 --location eastus --vm-sku Standard_D2s_v3 --instance-count 1')
```
Comment (MyronFanQiu) [inline src/azure-cli/azure/cli/command_modules/vm/tests/latest/test_vm_commands.py]: why do you add `--instance-count 1`?

File: src/azure-cli/azure/cli/command_modules/vm/tests/latest/test_vm_commands.py (hunk 3 above)
Comment (MyronFanQiu) [reply src/azure-cli/azure/cli/command_modules/vm/tests/latest/test_vm_commands.py]: It's another issue which should be fixed. Right?

File: src/azure-cli/azure/cli/command_modules/vm/custom.py (hunk 6 above)
Comment (qwordy) [reply src/azure-cli/azure/cli/command_modules/vm/custom.py]: Yes.

File: src/azure-cli/azure/cli/command_modules/vm/tests/latest/test_vm_commands.py (hunk 7 above)
Comment (qwordy) [reply src/azure-cli/azure/cli/command_modules/vm/tests/latest/test_vm_commands.py]: No enough quota

File: src/azure-cli/azure/cli/command_modules/vm/custom.py (hunk 1 above)
Comment (qwordy) [reply src/azure-cli/azure/cli/command_modules/vm/custom.py]: I will delete them and keep them in my private notebook.

File: src/azure-cli/azure/cli/command_modules/vm/tests/latest/test_vm_commands.py (hunk 3 above)
Comment (qwordy) [reply src/azure-cli/azure/cli/command_modules/vm/tests/latest/test_vm_commands.py]: Yes. I tried several other images and succeeded. When using ubuntults, timeout.

File: src/azure-cli/azure/cli/command_modules/vm/tests/latest/test_vm_commands.py (hunk 7 above)
Comment (qwordy) [reply src/azure-cli/azure/cli/command_modules/vm/tests/latest/test_vm_commands.py]: I tried 2, succeed.

**Pull request merged**